doc/cprover-manual/goto-harness.md

```diff
@@ -0,0 +1,263 @@
+* then we pass the input GOTO-binary and a name for the output GOTO binary.
+
+What comes of this is a GOTO binary roughly corresponding to the following C
+pseudocode:
+
+```C
+// modified_goto_binary.c
+
+#include <assert.h>
+
+int function_to_test(int some_argument)
+{
+  assert(some_argument == 0);
+  return some_argument;
+}
+
+void harness(void)
+{
+  int argument = nondet_int();
+  function_to_test(argument);
+}
+```
+
+You can pass that modified GOTO binary to `cbmc` and select `harness` as the
+entry function for analysis, or further modify it with `goto-harness` or
+`goto-instrument` or other tools.
+
+The example above demonstrates the most simple case, which is roughly the same
+as the entry point `cbmc` automically generates for functions. However, the
+`function-call` harness can also non-deterministically initialise global
+variables, array and struct elements, consider this more complicated example:
+
+```C
+// list_example.c
+#include <assert.h>
+#include <stdlib.h>
+
+typedef struct linked_list linked_list;
+struct linked_list {
+    int value;
+    linked_list *next;
+};
+
+linked_list *global_linked_list;
+
+/// initialize all values in the global
+/// list to 0
+void initialize_global(void)
+{
+  for(linked_list *ll = global_linked_list;
+    ll != NULL;
+    ll = ll->next)
+  {
+    ll->value = 0;
+  }
+}
+
+/// try to initialize all values in the linked list
+/// to 0 - but this version contains two bugs,
+/// as it won't work with nullpointer arguments
+/// and it will also not initialize the last element
+void initialize_other(linked_list *ll)
+{
+    do {
+        ll->value = 0;
+        ll = ll->next;
+    } while(ll->next != NULL);
+}
+
+void check_list(linked_list *list_parameter)
+{
+    assert(list_parameter != global_linked_list);
+    initialize_global();
+    initialize_other(list_parameter);
+    linked_list *global_cursor = global_linked_list;
+    linked_list *parameter_cursor = list_parameter;
+
+    // global list should be a prefix of the parameter now,
+    // or the other way round
+    while(global_cursor != NULL && parameter_cursor != NULL)
+    {
+        // this assertion should fail since we didn't
+        // initialize the last element of of the
+        // list parameter correctly
+        assert(global_cursor->value
+          == parameter_cursor->value);
+        global_cursor = global_cursor->next;
+        parameter_cursor = parameter_cursor->next;
+    }
+}
+```
+
+Now we'll try to find the bug in `check_list` by generating a harness for it.
+
+```sh
+$ goto-cc -o list_example.gb list_example.c
+$ goto-harness \
+   --harness-function-name harness \
+   --harness-type call-function \
+   --function check_list
+   --max-nondet-tree-depth 4 \
+   --min-null-tree-depth 1 \
+   --nondet-globals \
+   list_example.gb \
+   list_example_with_harness.gb
+$ cbmc --function harness list_example_with_harness.gb --unwind 20 --unwinding-assertions
```

Comment:
Do we really need to bound the unwinding?

Reply:
Yes, unfortunately symex can’t seem to figure out when to stop unwinding on its own here,